google-cloud-spanner/src/main/java/com/google/cloud/spanner/SimpleClientImpl.java

```diff
@@ -0,0 +1,131 @@
+
+  SimpleClientImpl(DatabaseClient client) {
+    this.client = client;
+  }
+
+  @Override
+  public ResultSet executeSqlQuery(String query) {
+    return executeSqlQuery(query, TimestampBound.strong());
+  }
+
+  @Override
+  public ResultSet executeSnapshotSqlQuery(String query) {
+    return executeSqlQuery(query, TimestampBound.ofExactStaleness(5L, TimeUnit.SECONDS));
+  }
+
+  @Override
+  public ResultSet executeSqlQuery(String query, TimestampBound timestampBound) {
+    return executeSqlQuery(query, null, timestampBound);
+  }
+
+  @Override
+  public ResultSet executeSqlQuery(
+      String query, Map<String, Value> args, TimestampBound timestampBound) {
+    try (ReadOnlyTransaction tx = client.singleUseReadOnlyTransaction(timestampBound)) {
+      Statement statement;
+      if (args != null && !args.isEmpty()) {
+        Statement.Builder builder = Statement.newBuilder(query);
+        for (Map.Entry<String, Value> entry : args.entrySet()) {
+          builder.bind(entry.getKey()).to(entry.getValue());
+        }
+        statement = builder.build();
+      } else {
+        statement = Statement.of(query);
+      }
+      return tx.executeQuery(statement);
+    }
+  }
+
+  @Override
+  public ResultSet executeSqlQuery(String query, Map<String, Value> args) {
+    return executeSqlQuery(query, args, TimestampBound.strong());
+  }
+
+  @Override
+  public ResultSet executeSnapshotSqlQuery(String query, Map<String, Value> args) {
+    return executeSqlQuery(query, args, TimestampBound.ofExactStaleness(5L, TimeUnit.SECONDS));
+  }
+
+  @Override
+  public long executeSqlWrite(String sql, Map<String, Value> args) {
+    Statement.Builder builder = Statement.newBuilder(sql);
+    if (args != null && !args.isEmpty()) {
+      for (Map.Entry<String, Value> entry : args.entrySet()) {
+        builder.bind(entry.getKey()).to(entry.getValue());
+      }
+    }
+    return client.executePartitionedUpdate(builder.build());
+  }
+
+  @Override
+  public boolean runTransaction(final List<Statement> statements) {
+    try {
+      boolean status =
+          client
+              .readWriteTransaction()
+              .run(
+                  new TransactionRunner.TransactionCallable<Boolean>() {
+                    @Override
+                    public Boolean run(TransactionContext transaction) {
+                      try {
+                        transaction.batchUpdate(statements);
+                        return true;
+                      } catch (Exception e) {
+                        logger.log(Level.WARNING, "Failed to complete transaction", e);
+                      }
+                      return false;
+                    }
+                  });
+    } catch (Exception e) {
+      throw SpannerExceptionFactory.newSpannerException(e);
+    }
+    return true;
+  }
+
+  @Override
+  public void runTransaction(final Statement readStatement, final OnReadHandler handler) {
+    client
+        .readWriteTransaction()
```

Comment:
You should use a single-use read-only transaction for this as well. Using a read/write transaction instead of a (single use) read-only transaction does not bring any advantages over using a read/write transaction, while it does involve more RPCs and (potentially) more locking.

Reply:
The use case here is read then write
so read write transaction is required.
see IT for example.